Source/Core/VideoBackends/Vulkan/VKScheduler.h

```diff
@@ -0,0 +1,158 @@
+// Copyright 2022 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#include <cstddef>
+#include <queue>
+
+#include "CommandBufferManager.h"
+#include "Common/Align.h"
+
+namespace Vulkan
+{
+class Scheduler
+{
+  class Command
+  {
+  public:
+    virtual ~Command() = default;
+
+    virtual void Execute(CommandBufferManager* cmdbuf) = 0;
+
+    Command* GetNext() const { return next; }
+
+    void SetNext(Command* next_) { next = next_; }
+
+  private:
+    Command* next = nullptr;
+  };
+
+  template <typename T>
+  class alignas(16) TypedCommand final : public Command
+  {
+  public:
+    explicit TypedCommand(T&& command_) : command{std::move(command_)} {}
+
+    ~TypedCommand() override = default;
+
+    TypedCommand(TypedCommand&&) = delete;
+
+    TypedCommand& operator=(TypedCommand&&) = delete;
+
+    void Execute(CommandBufferManager* cmdbuf) override { command(cmdbuf); }
+
+  private:
+    T command;
+  };
+
+  class CommandChunk final
+  {
+  public:
+    void ExecuteAll(CommandBufferManager* cmdbuf);
+
+    template <typename T>
+    bool Record(T& command)
+    {
+      using FuncType = TypedCommand<T>;
+      static_assert(sizeof(FuncType) < sizeof(data), "Lambda is too large");
+
+      command_offset = Common::AlignUp(command_offset, alignof(FuncType));
+      if (command_offset > sizeof(data) - sizeof(FuncType)) [[unlikely]]
+      {
+        return false;
+      }
+      Command* const current_last = last;
+      last = new (data.data() + command_offset) FuncType(std::move(command));
+
+      if (current_last) [[likely]]
+      {
+        current_last->SetNext(last);
+      }
+      else
+      {
+        first = last;
+      }
+      command_offset += sizeof(FuncType);
+      return true;
+    }
+
+    bool Empty() const { return command_offset == 0; }
+
+  private:
+    Command* first = nullptr;
+    Command* last = nullptr;
+
+    size_t command_offset = 0;
+    alignas(64) std::array<u8, 0x8000> data{};
+  };
+
+public:
+  Scheduler();
+  ~Scheduler();
+
+  bool Initialize();
+
+  void Flush();
+  void SyncWorker();
+
+  void Shutdown();
+
+  template <typename T>
+  void Record(T&& command)
+  {
+#ifdef VULKAN_DISABLE_THREADING
+    command(m_commandBufferManager.get());
+    return;
+#endif
+
+    if (m_chunk->Record(command)) [[likely]]
+      return;
+
+    Flush();
+    (void)m_chunk->Record(command);
```

Comment:
Would it be worth having an assertion / panic alert in case this still fails? Discarding error return values is the recipe for undebuggable issues in the future :-)

I understand that this shouldn't fail with how the code is currently implemented, but having the assertion would guarantee that future refactorings / improvements to the code preserve this assumption.

Reply:
Can't hurt, especially since it's not the hot path anyway.